ADR/0035-artifact-streams.md

```diff
@@ -0,0 +1,143 @@
+
+Date started: 2023-11-17
+Date accepted: 2023-02-01
+
+## Status
+
+Proposed
+
+Relates to:
+
+<!-- ... a lot ... -->
+
+Supercedes:
+
+<!-- ... a lot ... -->
+
+## Authors
+
+- Andrew M
+
+## Context
+
+Since the beginning of AppStudio, we have had a dependency on the [Component] and [Application] model that
+originated in the Hybrid Application Service. A [Component] is a single containerized artifact that has
+information on how to build and deploy recorded in a linked [Devfile]. Multiple [Components] can be grouped
+together into a single [Application] to indicate that the should be deployed together via GitOps.
+
+Along with the [gitops-service] which was decided to stop deploying in [ADR 32. Decoupling Deployment],
+the controllers' APIs and the UI for AppStudio have been tightly coupled with the concept of [Application]s
+and [Component]s. For example, 
+
+- PipelineRuns triggered by PipelinesAsCode produced a single image that is tied to a specific [Component]
+  which is assumed to be an OCI artifact.
+- The AppStudio UI (HAC) display PipelineRuns for all [Component]s in a single [Application] together.
+- [Component]s are tied to artifacts produced from a single a git repository and branch.
+
+This model presented multiple challenges including:
+
+- If a user wants to build a new branch of an already onboarded git repository, they would need to create
+  a new [Component] (with a workspace-unique name).
+- [Component]s can be members of only one [Application].
+- As the number of related [Component]s increases, it becomes much harder to visualize all of the currently
+  or previously running PipelineRuns since all [Component]s are included together. This has resulted in users
+  creating multiple [Application]s that are related to each other just to resolve the visual representation
+  challenges.
+- Since a PipelineRun produces a single [Component], it becomes harder to support build strategies that produce
+  multiple artifacts, i.e [ko].
+- [Component]s are only defined in the context of an [Application] and cannot be moved. When a user wants to
+  start building a Component, they need to either add it to a current [Application] and have its builds affecting
+  the IntegrationTestScenarios results of the generated Snapshots or the component needs to be created in its
+  own temporary   Application only to be deleted and recreated in the appropriate one once the builds are
+  properly functioning.
+
+## Decision
+
+We are going to abstract the concepts of the built artifacts and their relationships to each other outside of
+any other controller within AppStudio. If a controller wants to maintain complete separation from others (i.e.
+in order to be fully deployable on its own), it may use an adapter pattern to convert this abstraction to its
+own relevant internal data model.
+
+### ArtifactConfig
+
+An `ArtifactConfig` describes the metadata which is used to define an `Artifact`, an `ArtifactStream`, and
```

Comment:
I eventually think some example CRs might help visualize this.

Reply:
Yeah, I can work on these eventually. I can see how it would be great to get them written up before this ADR is approved for sure. I wanted to get discussion started before I was able to work on example CRs since those haven't been in my mind specifically.